Documentation/design-docs/host-tracing.md

```diff
@@ -0,0 +1,74 @@
+* `file_trace_listener` which routes tracing into a file. This listener is activated by setting the `COREHOST_TRACE=1` and also providing the full path to the trace file in `COREHOST_TRACEFILE=<path>` environment variable.
+
+Custom host can provide its own implementation of `host_trace_listener` by calling:
+``` C++
+void hostfxr_set_trace_listener(host_trace_listener* listener)
+```
+from the `hostfxr` component or
+``` C++
+void corehost_set_trace_listener(host_trace_listener* listener)
+```
+from the `hostpolicy` component.
+
+The functions behave exactly the same in both components. The `listener` parameter can be:
+* a pointer to an implementation of `host_trace_listener` which is then registered the only listener for all tracing.
+* `NULL` value which unregisters any previously registered listener. After this call tracing is disabled.
+
+Custom host can and should register the trace listener as the first thing it does with the respective host component to ensure that all tracing is routed to it.  
+
+Only one trace listener can be registered at any given time.  
+
+Registering custom trace listener or setting it to `NULL` overrides any tracing enabled by environment variables.
+
+The `hostfxr` component will propagate the trace listener to the `hostpolicy` component before it calls into it. So custom host only needs to register its trace listener with the `hostfxr` component and not both. The propagation of the trace listener is only done for the duration necessary after which it will be unregistered again. So custom host might need to register its own listener if it makes calls directly to `hostpolicy` on top of the calls to `hostfxr`.
+
+## Future investments
+### Trace content
+Currently the host components tend to trace a lot. The trace contains lot of interesting information but it's done in a very verbose way which is sometimes hard to navigate. Future investment should look at the common scenarios which are using the host tracing and optimize the trace output for those scenarios. This doesn't necessarily mean decrease the amount of tracing, but possibly introduce "summary sections" which would describe the end result decisions for certain scenarios.
+
+### Interaction with other diagnostics in the .NET Core
+The host tracing covers several areas which are interesting for diagnosing common failure patterns:
+* Finding and starting the runtime
+* Resolving frameworks
+* Resolving assemblies (and native libraries)
+
+Especially resolution of assemblies is tightly coupled with assembly binder behavior at runtime. As of now there's no correlation or cooperation between the host tracing and the runtime diagnostics (exceptions, logging) in the assembly binder. Future improvements might investigate ways to introduce some amount of cooperation to make the various diagnostic techniques easier to use together.
```

Comment:
+1 for better diagnostics in the CLR assembly binder, and whatever we can do to help with native runtime libraries as well.

Reply:
I'm working on that in a separate context. The plan is to improve exception messages and add binder logging into the runtime.